### PR TITLE
Add unit tests for `pkg/reconciler/common` | main branch | knative\pkg

### PR DESCRIPTION
**Pull request by nikhil-thomas**
# Changes

Add unit tests for `pkg/reconciler/common` package

Adds updates to dependencies needed for newly added unit tests in  `pkg/reconciler/common` package.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

    ```release-note
    NONE
    ```

### CODE REVIEW

**Review by vincent-pli — APPROVED** [review]

File: pkg/reconciler/common/install_test.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	return oldKODATAPATH, found
+}
+
+func resetKoDataPath(oldVal string, found bool) {
+	if !found {
+		os.Unsetenv(KoEnvKey)
+		return
+	}
+	os.Setenv(KoEnvKey, oldVal)
+}
```
Comment (vdemeester) [inline pkg/reconciler/common/install_test.go:45]: Why is this required ? :thinking:

File: pkg/reconciler/common/install_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/google/go-cmp/cmp"
+	mf "github.com/manifestival/manifestival"
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func setKoDataPath() (string, bool) {
+	oldKODATAPATH, found := os.LookupEnv("KO_DATA_PATH")
+	os.Setenv(KoEnvKey, filepath.Join(os.Getenv("GOPATH"), "src", "github.com", "tektoncd", "operator", "cmd", "operator", "kodata"))
```
Comment (vdemeester) [inline pkg/reconciler/common/install_test.go:35]: Any reason to not do `filepath.Join(os.Getenv("GOPATH"), "src/github.com/tektoncd/operator/cmd/operator/kodata")) ?
Also, doesn't this make the assumption that the repository is cloned in `$GOPATH/src/github.com/tektoncd/operator` ? That assumption might not be true :sweat_smile:

File: pkg/reconciler/common/unstructured.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+package common
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// NamespacedResource is an unstructured resource with the given apiVersion, kind, ns and name.
+func NamespacedResource(apiVersion, kind, ns, name string) *unstructured.Unstructured {
```
Comment (vdemeester) [inline pkg/reconciler/common/unstructured.go:24]: Any reason for this to be exported ?
Any reason for it to return a pointer ? (it seems like it is always *unreferencing* the point, used as `*NamespacedResource`)

File: pkg/reconciler/common/unstructured.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	resource := &unstructured.Unstructured{}
+	resource.SetAPIVersion(apiVersion)
+	resource.SetKind(kind)
+	resource.SetNamespace(ns)
+	resource.SetName(name)
+	return resource
+}
+
+// ClusterScopedResource is an unstructured resource with the given apiVersion, kind and name.
+func ClusterScopedResource(apiVersion, kind, name string) *unstructured.Unstructured {
```
Comment (vdemeester) [inline pkg/reconciler/common/unstructured.go:34]: Same as above

File: pkg/reconciler/common/install_test.go (hunk 2 above)
Comment (vdemeester) [reply pkg/reconciler/common/install_test.go:35]: Couldn't we use `testdata` for this (as below) ?

File: pkg/reconciler/common/releases_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+const (
+	VERSION          = "0.15.2"
+	TEKTON_PIPELINES = "testdata/kodata/tekton-pipeline/" + VERSION + "/release.notags.yaml"
+)
+
+func TestGetLatestRelease(t *testing.T) {
+	koPath := "testdata/kodata"
+
+	tests := []struct {
```
Comment (vdemeester) [inline pkg/reconciler/common/releases_test.go:35]: No need for table-test if there is only one case :stuck_out_tongue_closed_eyes:

File: pkg/reconciler/common/releases_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			version := latestRelease(test.component)
+			util.AssertEqual(t, version, test.expected)
+		})
+	}
+}
+
+func TestListReleases(t *testing.T) {
+	koPath := "testdata/kodata"
+
+	tests := []struct {
```
Comment (vdemeester) [inline pkg/reconciler/common/releases_test.go:60]: No need for table-test if there is only one case :stuck_out_tongue_closed_eyes:

**Review by vdemeester — COMMENTED** [review]
In general, we split tests into two functions : one for success cases, one for errors (that way no need for a convoluted error check & co in the test. It also allows to test the error content itself (`expectedError …`) in the error cases

Otherwise, few comments :angel:

File: pkg/reconciler/common/releases_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	"os"
+	"testing"
+
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	util "github.com/tektoncd/operator/pkg/reconciler/common/testing"
+)
+
+const (
+	VERSION          = "0.15.2"
+	TEKTON_PIPELINES = "testdata/kodata/tekton-pipeline/" + VERSION + "/release.notags.yaml"
```
Comment (savitaashture) [inline pkg/reconciler/common/releases_test.go:29]: Can we remove this as i don't see any usage

File: pkg/reconciler/common/releases_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+	tests := []struct {
+		name      string
+		component v1alpha1.TektonComponent
+		expected  []string
+	}{
+		{
+			name:      "tekton-pipeline",
+			component: &v1alpha1.TektonPipeline{},
+			expected:  []string{"0.15.2"},
```
Comment (savitaashture) [inline pkg/reconciler/common/releases_test.go:68]:
```suggestion
			expected:  []string{VERSION},
```

Comment (nikhil-thomas) [reply pkg/reconciler/common/releases_test.go:68]: Let me do something better. This specific test lists available manifests. So I shall add additional items in testdata.
so the `expected` becomes `[]string{"0.15.2", "0.14.3", "0.13.2"}`.

It is true that the operator will support only the latest version for install. However, the old manifest will have to be stored in the repository, as the operator can use it to make comparisons to identify obsolete resources, while performing an update.

File: pkg/reconciler/common/releases_test.go (hunk 7 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases_test.go:29]: yup :+1:

File: pkg/reconciler/common/releases_test.go (hunk 6 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases_test.go:60]: yagni :+1:

File: pkg/reconciler/common/releases_test.go (hunk 5 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases_test.go:35]: yagni :+1:  :stuck_out_tongue:

File: pkg/reconciler/common/install_test.go (hunk 1 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/install_test.go:45]: good old `cargo cult programming` :smile_cat: 
i have removed it :+1:

File: pkg/reconciler/common/install_test.go (hunk 2 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/install_test.go:35]: yup, `cargo cult again`.
removed :+1:

File: pkg/reconciler/common/unstructured.go (hunk 3 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/unstructured.go:24]: true, i have made the necessary changes. :+1:

File: pkg/reconciler/common/unstructured.go (hunk 4 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/unstructured.go:34]: true, i have made the necessary changes. :+1:

**Comment by nikhil-thomas** [comment]
> In general, we split tests into two functions : one for success cases, one for errors (that way no need for a convoluted error check & co in the test. It also allows to test the error content itself (`expectedError …`) in the error cases
> 
> Otherwise, few comments

I totally agree with you. Once this `main` branch catches up with `master`, I shall definitely refactor and improve the tests part by part. :+1:

**Review by vdemeester — APPROVED** [review]
/meow
